Zed crashes when dragging another app's tab over its window (Linux/Wayland)

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Dragging a tab from a Wayland-native app over Zed's main window makes it crash.
Apps running under Xwayland don't seem to trigger this behaviour.

Steps:
- Open Firefox / Chrome / Gnome Text Editor / Gnome Files (Nautilus).
- Create a new tab.
- Drag the tab (not the window) over Zed's window
- Zed crashes with the following log:

```
unknown object (4278190083), message error(ous)
Io error: Invalid argument (os error 22)
[2024-07-11T11:36:54+01:00 ERROR util] crates/gpui/src/platform/linux/wayland/client.rs:661: OtherError(IoError(Os { code: 22, kind: InvalidInput, message: "Invalid argument" }))
[2024-07-11T11:36:54+01:00 ERROR util] crates/lsp/src/lsp.rs:730: oneshot canceled
[2024-07-11T11:36:54+01:00 ERROR util] crates/lsp/src/lsp.rs:730: oneshot canceled
[2024-07-11T11:36:54+01:00 ERROR util] crates/lsp/src/lsp.rs:730: oneshot canceled
[2024-07-11T11:36:54+01:00 ERROR util] crates/lsp/src/lsp.rs:730: oneshot canceled
```

https://github.com/zed-industries/zed/assets/46565592/50591e58-09e2-4dfd-888d-15ada71695c6



### Environment

Zed: v0.142.6 (Zed)
OS: Linux Wayland arch unknown
Memory: 15.1 GiB
Architecture: x86_64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>
<!-- Click below this line and paste or drag-and-drop your log-->
2024-07-11T11:55:47+01:00 [INFO] ========== starting zed ==========
2024-07-11T11:55:47+01:00 [INFO] perform;
2024-07-11T11:55:47+01:00 [INFO] read_command;
2024-07-11T11:55:47+01:00 [INFO] read_command;
2024-07-11T11:55:47+01:00 [INFO] Opening main db
2024-07-11T11:55:47+01:00 [INFO] socket reader;
2024-07-11T11:55:47+01:00 [INFO] new;
2024-07-11T11:55:47+01:00 [INFO] keep_updated;
2024-07-11T11:55:47+01:00 [INFO] Using git binary path: None
2024-07-11T11:55:47+01:00 [INFO] extensions updated. loading 20, reloading 0, unloading 0
2024-07-11T11:55:47+01:00 [INFO] activate is not implemented on Linux, ignoring the call
2024-07-11T11:55:47+01:00 [INFO] Opening main db
2024-07-11T11:55:47+01:00 [INFO] perform;
2024-07-11T11:55:47+01:00 [INFO] read_command;
2024-07-11T11:55:47+01:00 [INFO] read_command;
2024-07-11T11:55:47+01:00 [INFO] socket reader;
2024-07-11T11:55:47+01:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-07-11T11:55:47+01:00 [WARN] Theme "Blackula" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-07-11T11:55:47+01:00 [INFO] Enabling Vulkan Portability
2024-07-11T11:55:47+01:00 [INFO] Enabling color space support
2024-07-11T11:55:47+01:00 [INFO] new;
2024-07-11T11:55:47+01:00 [INFO] keep_updated;
2024-07-11T11:55:47+01:00 [INFO] Adapter "Intel(R) Graphics (ADL GT2)"
2024-07-11T11:55:47+01:00 [INFO] No ray tracing extensions are supported
2024-07-11T11:55:47+01:00 [INFO] Using surface present mode MAILBOX
2024-07-11T11:55:47+01:00 [WARN] Unable to forbid exclusive full screen
2024-07-11T11:55:47+01:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-07-11T11:55:47+01:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-07-11T11:55:47+01:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Bare" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Basic" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Classic Dark" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Classic Light" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Default Dark" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Default Light" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode High Contrast Dark" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode High Contrast Light" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Low Key" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Midnight" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Presentation Dark" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Presentation Light" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Spartan" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Xcode Sunset" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [INFO] set status on client 0: Authenticating
2024-07-11T11:55:47+01:00 [INFO] perform;
2024-07-11T11:55:47+01:00 [INFO] read_command;
2024-07-11T11:55:47+01:00 [INFO] Opening main db
2024-07-11T11:55:47+01:00 [INFO] read_command;
2024-07-11T11:55:47+01:00 [INFO] socket reader;
2024-07-11T11:55:47+01:00 [INFO] Opening main db
2024-07-11T11:55:47+01:00 [INFO] new;
2024-07-11T11:55:47+01:00 [INFO] keep_updated;
2024-07-11T11:55:47+01:00 [WARN] Theme "Zedwaita Dark" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [WARN] Theme "Zedwaita Light" is using a deprecated style property: scrollbar_thumb.background. Use `scrollbar.thumb.background` instead.
2024-07-11T11:55:47+01:00 [INFO] Using surface present mode MAILBOX
2024-07-11T11:55:47+01:00 [WARN] Unable to forbid exclusive full screen
2024-07-11T11:55:47+01:00 [INFO] set status on client 162370: Connecting
2024-07-11T11:55:47+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:47+01:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-07-11T11:55:47+01:00 [INFO] starting language server "json-language-server", path: "/home/username/github/SomeRepo", id: 1
2024-07-11T11:55:47+01:00 [INFO] starting language server "package-version-server", path: "/home/username/github/SomeRepo", id: 2
2024-07-11T11:55:47+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:47+01:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-07-11T11:55:47+01:00 [INFO] Wayland does not support this API
2024-07-11T11:55:47+01:00 [INFO] Node runtime install_if_needed
2024-07-11T11:55:47+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Python: pyright
2024-07-11T11:55:47+01:00 [INFO] starting language server "pyright", path: "/home/username/github/SomeRepo", id: 3
2024-07-11T11:55:47+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:47+01:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Python: pyright
2024-07-11T11:55:47+01:00 [INFO] set environment variables from shell:/bin/bash, path:/usr/local/bin:/usr/bin:/usr/local/sbin:/opt/android-sdk/platform-tools:/opt/cuda/bin:/opt/cuda/nsight_compute:/opt/cuda/nsight_systems/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/username/.local/bin:/opt/android-sdk/platform-tools:/home/username/bin:/home/username/sources:/home/username/.local/bin:/home/username/.cargo/bin:/home/username/.local/bin
2024-07-11T11:55:47+01:00 [INFO] fetching latest version of language server "json-language-server"
2024-07-11T11:55:47+01:00 [INFO] fetching latest version of language server "package-version-server"
2024-07-11T11:55:47+01:00 [INFO] fetching latest version of language server "pyright"
2024-07-11T11:55:47+01:00 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
2024-07-11T11:55:47+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Terraform: 
2024-07-11T11:55:47+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Terraform: 
2024-07-11T11:55:47+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Terraform: 
2024-07-11T11:55:47+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:47+01:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Terraform: 
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Python: pyright
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Terraform: 
2024-07-11T11:55:47+01:00 [INFO] starting language servers for Terraform: 
2024-07-11T11:55:47+01:00 [INFO] downloading language server "package-version-server"
2024-07-11T11:55:47+01:00 [INFO] starting language server. binary path: "/home/username/.local/share/zed/languages/package-version-server/package-version-server-v0.0.3", working directory: "/home/username/github/SomeRepo", args: []
2024-07-11T11:55:47+01:00 [INFO] starting language server. binary path: "/home/username/.local/share/zed/node/node-v18.15.0-linux-x64/bin/node", working directory: "/", args: ["/home/username/.local/share/zed/copilot/copilot-v0.5.0/dist/agent.js", "--stdio"]
2024-07-11T11:55:47+01:00 [INFO] Node runtime install_if_needed
2024-07-11T11:55:47+01:00 [INFO] add connection to peer
2024-07-11T11:55:47+01:00 [INFO] add_connection;
2024-07-11T11:55:47+01:00 [INFO] waiting for server hello
2024-07-11T11:55:47+01:00 [INFO] got server hello
2024-07-11T11:55:47+01:00 [INFO] set status to connected (connection id: ConnectionId { owner_id: 0, id: 0 }, peer id: PeerId { owner_id: 451, id: 5887954 })
2024-07-11T11:55:47+01:00 [INFO] set status on client 162370: Connected { peer_id: PeerId { owner_id: 451, id: 5887954 }, connection_id: ConnectionId { owner_id: 0, id: 0 } }
2024-07-11T11:55:48+01:00 [INFO] Node runtime install_if_needed
2024-07-11T11:55:48+01:00 [INFO] Language server with id 0 sent unhandled notification LogMessage:
{
  "level": 0,
  "message": "[DEBUG] [agent] [2024-07-11T10:55:48.220Z] Agent service starting",
  "metadataStr": "[DEBUG] [agent] [2024-07-11T10:55:48.220Z]",
  "extra": [
    "Agent service starting"
  ]
}
2024-07-11T11:55:48+01:00 [INFO] Language server with id 0 sent unhandled notification client/registerCapability:
{
  "registrations": [
    {
      "id": "846d2b04-d0fe-475e-ac96-0767b8322666",
      "method": "workspace/didChangeWorkspaceFolders",
      "registerOptions": {}
    }
  ]
}
2024-07-11T11:55:48+01:00 [INFO] Language server with id 0 sent unhandled notification LogMessage:
{
  "level": 0,
  "message": "[DEBUG] [agent] [2024-07-11T10:55:48.230Z] Telemetry initialized",
  "metadataStr": "[DEBUG] [agent] [2024-07-11T10:55:48.230Z]",
  "extra": [
    "Telemetry initialized"
  ]
}
2024-07-11T11:55:48+01:00 [INFO] Initializing default prettier with plugins {}
2024-07-11T11:55:48+01:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-07-11T11:55:48+01:00 [INFO] starting language servers for Terraform: terraform-ls
2024-07-11T11:55:48+01:00 [INFO] starting language server "terraform-ls", path: "/home/username/github/SomeRepo", id: 4
2024-07-11T11:55:48+01:00 [INFO] starting language servers for Python: pyright
2024-07-11T11:55:48+01:00 [INFO] starting language servers for Terraform: terraform-ls
2024-07-11T11:55:48+01:00 [INFO] starting language servers for Terraform: terraform-ls
2024-07-11T11:55:48+01:00 [INFO] downloading language server "json-language-server"
2024-07-11T11:55:48+01:00 [INFO] Node runtime install_if_needed
2024-07-11T11:55:48+01:00 [INFO] starting language server. binary path: "/home/username/.local/share/zed/node/node-v18.15.0-linux-x64/bin/node", working directory: "/home/username/github/SomeRepo", args: ["/home/username/.local/share/zed/languages/json-language-server/node_modules/vscode-langservers-extracted/bin/vscode-json-language-server", "--stdio"]
2024-07-11T11:55:49+01:00 [INFO] starting language server. binary path: "/home/username/.local/share/zed/extensions/work/terraform/terraform-ls-v0.33.3/terraform-ls", working directory: "/home/username/github/SomeRepo", args: ["serve"]
2024-07-11T11:55:49+01:00 [INFO] downloading language server "pyright"
2024-07-11T11:55:49+01:00 [INFO] Node runtime install_if_needed
2024-07-11T11:55:49+01:00 [INFO] starting language server. binary path: "/home/username/.local/share/zed/node/node-v18.15.0-linux-x64/bin/node", working directory: "/home/username/github/SomeRepo", args: ["/home/username/.local/share/zed/languages/pyright/node_modules/pyright/langserver.index.js", "--stdio"]
2024-07-11T11:55:49+01:00 [INFO] Language server with id 3 sent unhandled notification window/logMessage:
{
  "type": 3,
  "message": "Pyright language server 1.1.371 starting"
}
2024-07-11T11:55:49+01:00 [INFO] Language server with id 3 sent unhandled notification window/logMessage:
{
  "type": 3,
  "message": "Server root directory: file:///home/username/.local/share/zed/languages/pyright/node_modules/pyright/dist"
}
2024-07-11T11:55:49+01:00 [INFO] Language server with id 3 sent unhandled notification window/logMessage:
{
  "type": 3,
  "message": "Starting service instance \"\""
}
2024-07-11T11:55:49+01:00 [INFO] language server 3: unregistered workspace/DidChangeWatchedFiles capability with id fce8d5e2-61fb-44db-9d86-f515d8731226
2024-07-11T11:55:51+01:00 [ERROR] crates/gpui/src/platform/linux/wayland/client.rs:661: OtherError(IoError(Os { code: 22, kind: InvalidInput, message: "Invalid argument" }))
2024-07-11T11:55:51+01:00 [ERROR] crates/lsp/src/lsp.rs:730: oneshot canceled
2024-07-11T11:55:51+01:00 [ERROR] crates/lsp/src/lsp.rs:730: oneshot canceled
2024-07-11T11:55:51+01:00 [ERROR] crates/lsp/src/lsp.rs:730: oneshot canceled
2024-07-11T11:55:51+01:00 [ERROR] crates/lsp/src/lsp.rs:730: oneshot canceled

<!-- Click above this line and paste or drag-and-drop your log--></pre></details>

--- a/crates/gpui/src/platform/linux/wayland/client.rs
+++ b/crates/gpui/src/platform/linux/wayland/client.rs
@@ -1803,7 +1803,6 @@ impl Dispatch<wl_data_device::WlDataDevice, ()> for WaylandClientStatePtr {
 
                             // Prevent dropping text from other programs.
                             if paths.is_empty() {
-                                data_offer.finish();
                                 data_offer.destroy();
                                 return;
                             }

